### ISSUE TITLE
Stan Interface Broken

### ISSUE THREAD

**Issue by cpfiffer** [TuringLang/Turing.jl#729]
It looks like currently the `sample` function when using the Stan interface requires the model to be a subtype of `Function`, but I think this went away after the compiler refactor.

Current:

```julia
function sample(  mf::T,
                  num_samples::Int,
                  num_warmup::Int,
                  save_warmup::Bool,
                  thin::Int,
                  ss::CmdStan.Sample
                ) where{T<:Function}
    return sample(mf, num_samples, num_warmup, save_warmup, thin, ss.adapt, ss.alg)
end
```

MWE:

```julia
using Turing, CmdStan
@model gdemo_d() = begin
  s ~ InverseGamma(2, 3)
  m ~ Normal(0, sqrt(s))
  1.5 ~ Normal(m, sqrt(s))
  2.0 ~ Normal(m, sqrt(s))
  return s, m
end

gdemo_default = gdemo_d()
chn = sample(gdemo_d(), 2000, 1000, false, 1,
    CmdStan.Adapt(), CmdStan.Hmc())
```

Error: 

```julia
ERROR: LoadError: MethodError: no method matching sample(::Turing.Model{Tuple{:s,:m},Tuple{},getfield(Main, Symbol("###inner_function#5455#1735")),NamedTuple{(),Tuple{}},NamedTuple{(),Tuple{}}}, ::Int64, ::Int64, ::Bool, ::Int64, ::CmdStan.Adapt, ::CmdStan.Hmc)
Closest candidates are:
  sample(::T<:Function, ::Int64, ::Int64, ::Bool, ::Int64, ::CmdStan.Adapt, ::CmdStan.Hmc) where T<:Function at /home/cameron/.julia/dev/Turing/src/utilities/stan-interface.jl:35
  sample(::T<:Function, ::Int64, ::Int64, ::Bool, ::Int64, ::Sample) where T<:Function at /home/cameron/.julia/dev/Turing/src/utilities/stan-interface.jl:24
  sample(::AbstractRNG, ::AbstractArray{T,N} where N, ::Integer; replace, ordered) where T at /home/cameron/.julia/packages/StatsBase/XePnn/src/sampling.jl:373
  ...
```

**Comment by xukai92** [comment]
Can we change `T<:Function` to `Model`?

**Comment by cpfiffer** [comment]
Probably? I can assemble a PR based on the other samplers and see if it works

**Comment by yebai** [comment]
Fixed by 3abbb83
